async/operations.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include "infra.hpp"
+
+enum class UnaryType {
+    Relu,
+};
+
+struct UnaryDeviceOperation {
+    struct operation_attributes_t {
+        UnaryType unary_type;
+        Dtype dtype;
+    };
+
+    struct tensor_args_t {
+        Tensor input_tensor;
+    };
+
+    using tensor_return_value_t = Tensor;
+
+    // select_program_factory
+    // validate on program cache miss
+    // validate on program cache hit
+    // compute output shapes
+
+    static tensor_return_value_t create_output_tensors(
+        const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+        return Tensor{
+            tensor_args.input_tensor.shape, attributes.dtype, DeviceStorage{get_device(tensor_args.input_tensor)}};
+    }
+
+    static Program create_program(
+        const operation_attributes_t& attributes,
+        const tensor_args_t& tensor_args,
+        tensor_return_value_t& tensor_return_value) {
+        return Program{};
+    }
+};
+
+// Primitive Operation
+template <UnaryType type>
+struct UnaryOperation {
+    using device_operation_t = UnaryDeviceOperation;
+
+    static auto map_args_to_device_operation(const Tensor& input_tensor, Dtype dtype = Dtype::bfloat16) {
+        return std::make_tuple(
+            device_operation_t::operation_attributes_t{.unary_type = type, .dtype = dtype},
+            device_operation_t::tensor_args_t{.input_tensor = input_tensor});
+    }
+};
+
+constexpr auto relu = register_operation<"relu", UnaryOperation<UnaryType::Relu>>();
```

Comment:
In a header, globally scoped variables should be `inline constexpr` instead of `constexpr` to avoid potential ODR violations.